### PR TITLE
Support parquet file for log_table

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/11922?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11922/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11922
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Support log_table to parquet file

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -2360,11 +2368,11 @@ def log_table(
         import pandas as pd
 
         self._check_artifact_file_string(artifact_file)
-        if not artifact_file.endswith(".json"):
+        if not artifact_file.endswith(".json") and not artifact_file.endswith(".parquet"):
```
Comment (harupy) [inline mlflow/tracking/client.py:2371]: /ai simplify this line

File: tests/tracking/test_tracking.py [hunk 2]
```diff
@@ -850,19 +850,29 @@ def test_search_runs_multiple_experiments():
     assert len(MlflowClient().search_runs(experiment_ids, "metrics.m_3 < 4", ViewType.ALL)) == 1
 
 
+def read_data(artifact_path):
+    import pandas as pd
+
+    if artifact_path.endswith("json"):
+        return pd.read_json(artifact_path, orient="split")
+    if artifact_path.endswith("parquet"):
```
Comment (harupy) [inline tests/tracking/test_tracking.py:858]:
```suggestion
    if artifact_path.endswith(".json"):
        return pd.read_json(artifact_path, orient="split")
    if artifact_path.endswith(".parquet"):
```

File: tests/tracking/test_tracking.py [hunk 3]
```diff
@@ -850,19 +850,29 @@ def test_search_runs_multiple_experiments():
...
 
 
+def read_data(artifact_path):
+    import pandas as pd
+
+    if artifact_path.endswith(".json"):
+        return pd.read_json(artifact_path, orient="split")
+    if artifact_path.endswith(".parquet"):
+        return pd.read_parquet(artifact_path)
+
```
Comment (harupy) [inline tests/tracking/test_tracking.py:860]: /ai throw an error if `artifact_path` is an unexpected file type.

Comment (harupy) [reply tests/tracking/test_tracking.py:860]: /ai Nice. Let's include `artifact_path` in the error message.

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/11922/files#r1591781422 is addressed!

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -2360,11 +2369,11 @@ def log_table(
...
 
         self._check_artifact_file_string(artifact_file)
-        if not artifact_file.endswith(".json"):
+        if not artifact_file.endswith((".json", ".parquet")):
             raise ValueError(
                 f"The provided artifact file '{artifact_file}' does not have "
-                "the required '.json' extension. Please ensure the file you are trying "
-                "to log as a table is a JSON file with the '.json' extension. "
+                "the required '.json' or '.parquet' extension. Please ensure the file you are "
+                "trying to log as a table is valid with either '.json' or '.parquet' extension."
```
Comment (WeichenXu123) [inline mlflow/tracking/client.py:2376]:
```suggestion
                "trying to log as a table has a file name with either '.json' or '.parquet' extension."
```

File: mlflow/tracking/client.py [hunk 5]
```diff
@@ -2286,6 +2286,15 @@ def _check_artifact_file_string(self, artifact_file: str):
             if char in artifact_file:
                 raise ValueError(f"The artifact_file contains forbidden character: {char}")
 
+    def _read_from_file(self, artifact_path):
```
Comment (WeichenXu123) [inline mlflow/tracking/client.py:2289]: can it call `read_data` function ?

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

**Pull request merged**